### PR TITLE
LINK-2024 | improve test coverage

### PR DESCRIPTION
**Pull request by jorilindell**
## Description
- Adds tests for the components and the functions which were not fully tested.
- Removes some unused functions.
- Also refactors scrollToFirstError and getFocusableFieldId functions to avoid duplicated code. Also fine-tuned all form specific e.g. getFocusableRegistrationFieldId functions to make that focus is moved to the correct field after validation error

## Closes
[LINK-2024](https://helsinkisolutionoffice.atlassian.net/browse/LINK-2024)

[LINK-2024]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-2024?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by terovirtanen** [comment]
**LINKEDCOMPONENTS-UI branch is deployed to platta: https://linkedcomponents-pr414-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __failed__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
**LINKEDCOMPONENTS-UI branch is deployed to platta: https://linkedcomponents-pr414-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __failed__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**LINKEDCOMPONENTS-UI branch is deployed to platta: https://linkedcomponents-pr414-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :laughing::tada::tada::tada:

File: src/domain/registration/utils.ts [hunk 1]
```diff
@@ -678,3 +680,28 @@ export const exportSignupsAsExcel = ({
       addNotification({ type: 'error', label: error.message });
     });
 };
+
+export const getPriceGroupOptionsForPriceGroup = ({
```
Comment (harriris-vincit) [inline src/domain/registration/utils.ts:684]: This could maybe be in a separate refactoring commit together with the changes in `src/domain/registration/formSections/priceGroups/priceGroup/PriceGroup.tsx` and in `src/domain/registration/__tests__/utils.test.ts`

File: src/utils/getUserDisplayName.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { User } from 'oidc-client-ts';
+
+import skipFalsyType from './skipFalsyType';
+
+const getUserDisplayName = ({
```
Comment (harriris-vincit) [inline src/utils/getUserDisplayName.ts:5]: Here is another one where all changes related to the introduction of this util function could be in their own refactoring commit.

File: src/utils/validationUtils.ts [hunk 3]
```diff
@@ -279,24 +279,99 @@ export const showFormErrors = <Values = unknown>({
...
+
+export type ErrorFieldAndType = {
+  fieldId: string;
+  type: ErrorFieldType;
+};
+
+const isInFieldList = (fieldList: FieldList, fieldName: string): boolean =>
+  Boolean(fieldList.find((field) => new RegExp(field).test(fieldName)));
+
+export const getFocusableFieldId = (
```
Comment (harriris-vincit) [inline src/utils/validationUtils.ts:308]: This one could use a bit of refactoring to get rid of the `if-else if` structure (**code not tested**):
```
const fieldMappings: Record<keyof FieldLists, { suffix: string, type: ErrorFieldType }> = {
  comboboxFields: { suffix: '-input', type: 'combobox' },
  selectFields: { suffix: '-toggle-button', type: 'select' },
  textEditorFields: { suffix: '-text-editor', type: 'textEditor' },
  checkboxGroupFields: { suffix: '', type: 'checkboxGroup' },
  arrayFields: { suffix: '-error', type: 'array' },
};

export const getFocusableFieldId = (
  fieldName: string,
  fieldLists: FieldLists
): ErrorFieldAndType => {
  for (const [fieldListKey, { suffix, type }] of Object.entries(fieldMappings)) {
    if (isInFieldList(fieldLists[fieldListKey as keyof FieldLists], fieldName)) {
      return { fieldId: `${fieldName}${suffix}`, type };
    }
  }

  return { fieldId: fieldName, type: 'default' };
};
```

File: src/common/components/formFields/multiLanguageField/MultiLanguageField.tsx [hunk 4]
```diff
@@ -29,21 +27,22 @@ const MultiLanguageField: React.FC<Props> = ({
   ...rest
 }) => {
   const { t } = useTranslation();
-  if (!languages.length) return null;
 
   return (
     <div className={styles.multiLanguageField}>
-      {ORDERED_LE_DATA_LANGUAGES.map((language) => {
-        const langText = lowerCaseFirstLetter(t(`form.inLanguage.${language}`));
+      {ORDERED_LE_DATA_LANGUAGES.filter((l) => languages.includes(l)).map(
```
Comment (harriris-vincit) [inline src/common/components/formFields/multiLanguageField/MultiLanguageField.tsx:33]: Should maybe be in it's own commit?

File: src/domain/event/hooks/useUpdateImageIfNeeded.tsx [hunk 5]
```diff
@@ -36,21 +40,28 @@ const useUpdateImageIfNeeded = (): UpdateImageIfNeededState => {
   const { user } = useUser();
   const [updateImage] = useUpdateImageMutation();
 
-  const cleanAfterUpdate = async () => {
```
Comment (harriris-vincit) [inline src/domain/event/hooks/useUpdateImageIfNeeded.tsx:39]: Changes in this file could probably be in their own commit.

**Comment by terovirtanen** [comment]
**LINKEDCOMPONENTS-UI branch is deployed to platta: https://linkedcomponents-pr414-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :laughing::tada::tada::tada:

File: src/utils/validationUtils.ts (hunk 3 above)
Comment (harriris-vincit) [reply src/utils/validationUtils.ts:308]: Though this can be considered an optional change as this function has existed before with the `if-else if` structure.

Comment (jorilindell) [reply src/utils/validationUtils.ts:319]: Really nice suggestion. Changed as suggested

File: src/domain/event/hooks/useUpdateImageIfNeeded.tsx (hunk 5 above)
Comment (jorilindell) [reply src/domain/event/hooks/useUpdateImageIfNeeded.tsx:39]: These changes are now moved to a separate commit

**Comment by terovirtanen** [comment]
**LINKEDCOMPONENTS-UI branch is deployed to platta: https://linkedcomponents-pr414-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Review by harriris-vincit — APPROVED** [review]
Huge amount of changes, but LGTM now :+1: Tested multiple different pages and they still seemed to function as they should.

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://linkedcomponents-pr414-ui.dev.hel.ninja :laughing::tada::tada::tada:
